File: solutions/ocaml/079-gas-station.ml
(*
  Problem 79: Gas Station (LeetCode 134)
  Difficulty: Med
  Language: OCaml
*)
let can_complete_circuit gas cost =
  let diffs = List.map2 (-) gas cost in
  let (total, _, start) = List.fold_left (fun (tot, tank, s) (d, i) ->
    let tot' = tot + d and tank' = tank + d in
    if tank' < 0 then (tot', 0, i + 1) else (tot', tank', s))
    (0, 0, 0) (List.mapi (fun i d -> (d, i)) diffs) in
  if total >= 0 then start else -1

File: solutions/scala/079-gas-station.scala
/**
 * Problem 79: Gas Station (LeetCode 134)
 * Difficulty: Med
 * Language: Scala
 */
def canCompleteCircuit(gas: List[Int], cost: List[Int]): Int = {
  val (total, _, start) = gas.zip(cost).zipWithIndex.foldLeft((0, 0, 0)) {
    case ((tot, tank, s), ((g, c), i)) =>
      val d = g - c; val (t2, tk2) = (tot + d, tank + d)
      if (tk2 < 0) (t2, 0, i + 1) else (t2, tk2, s)
  }
  if (total >= 0) start else -1
}
